internal/pkg/consensus/election.go

```diff
@@ -92,23 +99,51 @@ func (em ElectionMachine) VerifyPoSt(ep verification.PoStVerifier, allSectorInfo
 
 // TicketMachine uses a VRF and VDF to generate deterministic, unpredictable
 // and time delayed tickets and validates these tickets.
-type TicketMachine struct{}
+type TicketMachine struct {
+	chain ChainRandomness
+}
 
-// NextTicket creates a new ticket from a parent ticket by running a verifiable
-// randomness function on the parent.
-func (tm TicketMachine) NextTicket(parent block.Ticket, signerAddr address.Address, signer types.Signer) (block.Ticket, error) {
-	vrfPi, err := signer.SignBytes(parent.VRFProof[:], signerAddr)
+func NewTicketMachine(chain ChainRandomness) *TicketMachine {
+	return &TicketMachine{chain: chain}
+}
+
+// MakeTicket creates a new ticket from a chain and target epoch by running a verifiable
+// randomness function on the prior ticket.
+func (tm TicketMachine) MakeTicket(ctx context.Context, base block.TipSetKey, epoch abi.ChainEpoch, miner address.Address, worker address.Address, signer types.Signer) (block.Ticket, error) {
+	vrfProof, err := computeVrf(ctx, tm.chain, base, epoch, acrypto.DomainSeparationTag_TicketProduction, miner, signer, worker)
+	return block.Ticket{
+		VRFProof: vrfProof,
+	}, err
+}
+
+// IsValidTicket verifies that the ticket's proof of randomness is valid with respect to its parent.
+func (tm TicketMachine) IsValidTicket(ctx context.Context, base block.TipSetKey,
+	epoch abi.ChainEpoch, miner address.Address, worker address.Address, ticket block.Ticket) error {
+	entropy, err := encoding.Encode(miner)
+	if err != nil {
+		return errors.Wrapf(err, "failed to encode entropy")
+	}
+	randomness, err := tm.chain.SampleChainRandomness(ctx, base, acrypto.DomainSeparationTag_TicketProduction, epoch, entropy)
 	if err != nil {
-		return block.Ticket{}, err
+		return errors.Wrap(err, "failed to generate epost randomness")
 	}
 
-	return block.Ticket{
-		VRFProof: vrfPi.Data,
-	}, nil
+	return crypto.ValidateBlsSignature(randomness, worker, ticket.VRFProof)
 }
 
-// ValidateTicket verifies that the ticket's proof of randomness and delay are
-// valid with respect to its parent.
-func (tm TicketMachine) ValidateTicket(parent, ticket block.Ticket, signerAddr address.Address) error {
-	return crypto.ValidateBlsSignature(parent.VRFProof[:], signerAddr, ticket.VRFProof)
+func computeVrf(ctx context.Context, chain ChainRandomness, base block.TipSetKey, epoch abi.ChainEpoch, tag acrypto.DomainSeparationTag,
```

Comment:
nit: I'd prefer the name to reflect that this does randomness sampling work along with computing a VRF.  But nbd as the arguments mostly tell the story.